### ISSUE TITLE
support for azure arm template parameters of type securestring

### ISSUE THREAD

**Issue by jritsema** [hashicorp/terraform#8331]
It doesn't look like the terraform `azurerm` provider currently supports arm template parameters of type `securestring` in the `azurerm_template_deployment` resource.

You can see an example of this [here](https://azure.microsoft.com/en-us/documentation/articles/resource-manager-keyvault-parameter/).

Looks like the code always outputs this structure...

```json
"parameters": {
  "myAdminUsername": {
    "value": "MyAdministrator"
  }
}
```

rather than what's needed for `securestring`...

```json
"parameters": {
  "myAdminPassword": {
    "reference": {
      "keyVault": {
        "id": "..."
      },
      "secretName": "..."
    }
  }
}
```

**Comment by jritsema** [comment]
Has anyone considered adding a `parameters_body` property that would allow you to pass a parameters file reference?

```hcl
resource "azurerm_template_deployment" "deployment" {
  template_body = "${file("azuredeploy.json")}"
  parameters_body = "${file("azuredeploy.parameters.json")}"
```

That would actually solve this issue.  Just a thought.

**Comment by colemickens** [comment]
@jritsema, I'm confused, where is it documented that `securestring` turns into a KeyVault reference? And where does that transformation happen?

**Comment by colemickens** [comment]
I found an example of what you're talking about and I think there's some confusion at play.

Using `securestring` doesn't make anything magically use KeyVault. It just happens to be the case that in your template, you are able to reference a KeyVault secret and have it's value used in place of a normal parameter filed (and it should only be used with `securestring` to prevent it from appearing in deployment logs and the persisted deployment model after submitting it).

So in this case, I guess you're asking for Terraform to automatically upload secrets into KeyVault and then inject references for certain parameters in your template deployment. It's an interesting idea, but not trivial to implement, especially given the lack of KeyVault support in Terraform today.

**Comment by jritsema** [comment]
Right on @colemickens.  I did conflate `securestring` and `keyvalut` when they do appear to be different concepts.  

So really I was trying to solve for was having terraform work with the magical `reference` to the key vault value in a parameters file.  I ended up solving this by just moving the bulk of my arm template into a linked template and embedding the key vault references in the parent template.

I'll go ahead and close this.
